ast/walk.go

```diff
@@ -72,6 +72,112 @@ func Walk(expr Expression, visitor func(Expression) bool) bool {
 	panic(fmt.Sprintf("Unexpected AST expression of type %T: %[1]v", expr))
 }
 
+// WalkAndFindParentUsageInvalid is like Walk() but it returns an error if an invalid
+// parent usage is found.
+// This function is exclusively used to walk the body of a custom defined function.
+// The usage of a parent operator is *not* allowed in the body of a custom defined function
+// if it reaches out of that function's scope. It's valid within a filter or a projection, because
+// we create NewNestedScope for those.
+func WalkAndFindParentUsageInvalid(count int, expr Expression) (int, error) {
+	if expr == nil {
+		return count, nil
+	}
+	switch e := expr.(type) {
+	case *Parent:
+		count--
+		if count <= 0 {
+			return count, fmt.Errorf("Parent usage is invalid")
+		}
+		return count, nil
+	case *Everything, *This, *Ellipsis, *StringLiteral,
+		*FloatLiteral, *BooleanLiteral, *IntegerLiteral, *NullLiteral, *Attribute,
+		*ArrayTraversal, *Param, *FunctionParam:
+		return count, nil
+	case *Constraint:
+		return WalkAndFindParentUsageInvalid(count, e.Expression)
+	case *Subscript:
+		return WalkAndFindParentUsageInvalid(count, e.Value)
+	case *Range:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.Start); err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.End)
+	case *BinaryOperator:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return c, err
+		}
```

Comment:
In general for these I'd expect the error case to look like this (unless it's intentional that we return both a count and an error at the same time ).

```suggestion
		if _, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
			return 0, err
		}
```